### PR TITLE
[SPARK-44370][CONNECT] Migrate Buf remote generation alpha to remote plugins

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
Buf unsupported remote generation alpha at now. Please refer https://buf.build/docs/migration-guides/migrate-remote-generation-alpha/ . We should migrate Buf remote generation alpha to remote plugins by follow the guide.

The CI also broken by this reason.

### Why are the changes needed?
Migrate Buf remote generation alpha to remote plugins because remote generation alpha features have been sunset.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
exist test

### CODE REVIEW

**Review by HyukjinKwon — APPROVED** [review]
Thanks for taking a look.

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

File: connector/connect/common/src/main/buf.gen.yaml [hunk 1]
```diff
@@ -16,18 +16,18 @@
 #
 version: v1
 plugins:
-  - remote: buf.build/protocolbuffers/plugins/cpp:v3.20.0-1
+  - plugin: buf.build/protocolbuffers/cpp:v21.7
     out: gen/proto/cpp
-  - remote: buf.build/protocolbuffers/plugins/csharp:v3.20.0-1
+  - plugin: buf.build/protocolbuffers/csharp:v21.7
```
Comment (LuciferYang) [inline connector/connect/common/src/main/buf.gen.yaml:21]: also cc @zhengruifeng @panbingkun

**Comment by zhengruifeng** [comment]
@LuciferYang @panbingkun @Hisoka-X @HyukjinKwon 

do you know what cause this large changes in generated codes? the upgrade to v1.23.1? or this migration?

File: connector/connect/common/src/main/buf.gen.yaml [hunk 2]
```diff
@@ -16,18 +16,18 @@
...
+  - plugin: buf.build/protocolbuffers/java:v21.7
     out: gen/proto/java
   - plugin: buf.build/grpc/ruby:v1.56.0
     out: gen/proto/ruby
-  - remote: buf.build/protocolbuffers/plugins/ruby:v21.2.0-1
+  - plugin: buf.build/protocolbuffers/ruby:v21.7
     out: gen/proto/ruby
    # Building the Python build and building the mypy interfaces.
-  - remote: buf.build/protocolbuffers/plugins/python:v3.19.3-1
+  - plugin: buf.build/protocolbuffers/python:v21.7
```
Comment (panbingkun) [inline connector/connect/common/src/main/buf.gen.yaml:30]: https://buf.build/protocolbuffers/python
plugin: buf.build/protocolbuffers/python:v23.4
It seems that this version is the latest.

**Comment by LuciferYang** [comment]
Can we try revert the version of buf 1.23.1 and only updating the `remote-plugins`?  If don't need to change so much code, I think we can revert buf version before code freeze

File: connector/connect/common/src/main/buf.gen.yaml [hunk 3]
```diff
@@ -16,18 +16,18 @@
 #
 version: v1
 plugins:
-  - remote: buf.build/protocolbuffers/plugins/cpp:v3.20.0-1
+  - plugin: buf.build/protocolbuffers/cpp:v21.7
```
[the remark this answers is not in the document]
Comment (panbingkun) [reply connector/connect/common/src/main/buf.gen.yaml:19]: Maybe as follow:
- plugin: buf.build/grpc/cpp:v1.56.0
    out: gen/proto/cpp
- plugin: buf.build/protoco
 out: gen/proto/cpp

Its writing is consistent with the original plugin's Ruby.

**Comment by panbingkun** [comment]
Let's try revert buf.

**Comment by zhengruifeng** [comment]
but i think we have pinned the version of protobuf  https://github.com/apache/spark/blob/d7bc6f5c7efd9dcb1e460657447d1a9ea8f03f62/.github/workflows/build_and_test.yml#L637

**Comment by zhengruifeng** [comment]
> 

yeah, this failure actually started before we upgrade buf

**Comment by zhengruifeng** [comment]
if we can not avoid codegen change, then i am fine to upgrade buf to latest

**Comment by panbingkun** [comment]
> > From the test results, it is not a problem with the Buf version. image
> 
> So even using buf 1.20 cannot avoid a lot of code changes, right? @panbingkun

Yeah.
image

**Review by zhengruifeng — APPROVED** [review]

**Review by panbingkun — APPROVED** [review]

**Comment by grundprinzip** [comment]
We have to be careful that the remotely generated protobuf code matches our protobuf version locally because otherwise we will run into compatibility issues. Protobuf had some changes which might cause the large diff in this PR.

What is the diff if we use the 3.14 version of protobuf?

**Comment by grundprinzip** [comment]
Do we know what the changes in the generated code are?

**Comment by grundprinzip** [comment]
I think using the 21.7 version might not be very risky, but we should make sure that we stay consistent in our protobuf usage and we need to check if this should apply as well to upgrading the java version.

**Comment by LuciferYang** [comment]
Yeah, Java use `sbt-protoc` or `protobuf-maven-plugin`, not `buf`

**Comment by panbingkun** [comment]
I guess it might be related to this:
https://buf.build/blog/breaking-change-governance/
image

**Comment by grundprinzip** [comment]
TBH I don't know how exactly to fix this problem. The buf remote plugins don't offer the protobuf version we need, so maybe the fix is to switch from remote to local plugins.

**Comment by grundprinzip** [comment]
My suggestion would be to merge the proposed version with v21.7 and assuming that the integration test pass should be still compatible. 

This should unblock any new proto changes.

**Comment by panbingkun** [comment]
@Hisoka-X Local plugins:
https://buf.build/docs/generate/usage/#3.-use-local-plugins

**Review by grundprinzip — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**